public/app/plugins/datasource/loki/shardQuerySplitting.ts

```diff
@@ -0,0 +1,351 @@
+ *   . groupShardRequests() returns an array of shards from cycle to cycle + groupSize.
+ *   . interpolateShardingSelector() will update the stream selector with the shard numbers in the current group.
+ *   . After query execution:
+ *     - If the response is successful:
+ *       . It will add new data to the response with combineResponses()
+ *       . Using the data and meta data of the response, updateGroupSizeFromResponse() will increase or decrease the group size.
+ *       . nextRequest() will use the current cycle and group size to determine the next request or complete execution with done().
+ *     - If the response is unsuccessful:
+ *       . If the response is not a query error, and the group size bigger than 1, it will decrease the group size.
+ *       . If the group size is already 1, it will retry the request up to 4 times.
+ *       . If there are retry attempts, it will retry the current cycle, or else stop querying.
+ * - Once all request groups have been executed, it will be done().
+ */
+
+export function runShardSplitQuery(datasource: LokiDatasource, request: DataQueryRequest<LokiQuery>) {
+  const queries = datasource
+    .interpolateVariablesInQueries(request.targets, request.scopedVars)
+    .filter((query) => query.expr)
+    .filter((query) => !query.hide);
+
+  return splitQueriesByStreamShard(datasource, request, queries);
+}
+
+function splitQueriesByStreamShard(
+  datasource: LokiDatasource,
+  request: DataQueryRequest<LokiQuery>,
+  splittingTargets: LokiQuery[]
+) {
+  let shouldStop = false;
+  let mergedResponse: DataQueryResponse = { data: [], state: LoadingState.Streaming, key: uuidv4() };
+  let subquerySubscription: Subscription | null = null;
+  let retriesMap = new Map<string, number>();
+  let retryTimer: ReturnType<typeof setTimeout> | null = null;
+
+  const runNextRequest = (subscriber: Subscriber<DataQueryResponse>, group: number, groups: ShardedQueryGroup[]) => {
+    let nextGroupSize = groups[group].groupSize;
+    const { shards, groupSize, cycle } = groups[group];
+    let retrying = false;
+
+    if (subquerySubscription != null) {
+      subquerySubscription.unsubscribe();
+      subquerySubscription = null;
+    }
+
+    if (shouldStop) {
+      subscriber.complete();
+      return;
+    }
+
+    const done = () => {
+      mergedResponse.state = LoadingState.Done;
+      subscriber.next(mergedResponse);
+      subscriber.complete();
+    };
+
+    const nextRequest = () => {
+      const nextGroup =
+        groups[group + 1] && groupHasPendingRequests(groups[group + 1])
+          ? groups[group + 1]
+          : groups.find((shardGroup) => groupHasPendingRequests(shardGroup));
+
+      if (nextGroup === undefined) {
+        done();
+        return;
+      }
+      groups[group].groupSize = nextGroupSize;
+      runNextRequest(subscriber, groups.indexOf(nextGroup), groups);
+    };
+
+    const retry = (errorResponse?: DataQueryResponse) => {
```

Comment:
This is duplicated code from `public/app/plugins/datasource/loki/querySplitting.ts`. If this is needed, can we reuse that code? Otherwise, I'm also fine not using any retry logic for now.

Reply:
Now both share isRetriableError() from responseUtils.